infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCProvisioner.java

```diff
@@ -0,0 +1,214 @@
+  }
+
+  /**
+   * Groups list of given PVCs by volume name. The result may be used for easy accessing to PVCs by
+   * Che Volume name.
+   */
+  private Map<String, PersistentVolumeClaim> groupByVolumeName(
+      Collection<PersistentVolumeClaim> pvcs) {
+    final Map<String, PersistentVolumeClaim> grouped = new HashMap<>();
+    for (PersistentVolumeClaim pvc : pvcs) {
+      final ObjectMeta metadata = pvc.getMetadata();
+      final String volumeName;
+      if (metadata.getLabels() != null
+          && (volumeName = metadata.getLabels().get(CHE_VOLUME_NAME_LABEL)) != null) {
+        grouped.put(volumeName, pvc);
+      } else {
+        grouped.put(metadata.getName(), pvc);
+        putLabel(metadata, CHE_VOLUME_NAME_LABEL, metadata.getName());
+      }
+    }
+    return grouped;
+  }
+
+  private void addMachineVolumes(
+      String workspaceId,
+      KubernetesEnvironment k8sEnv,
+      Map<String, PersistentVolumeClaim> volumeName2PVC,
+      PodData pod,
+      Container container,
+      Map<String, Volume> volumes) {
+    if (volumes.isEmpty()) {
+      return;
+    }
+
+    for (Entry<String, Volume> volumeEntry : volumes.entrySet()) {
+      final String volumePath = volumeEntry.getValue().getPath();
+      final String volumeName =
+          LOGS_VOLUME_NAME.equals(volumeEntry.getKey())
+              ? volumeEntry.getKey() + '-' + pod.getMetadata().getName()
+              : volumeEntry.getKey();
+      final PersistentVolumeClaim pvc;
+      // checks whether PVC for given workspace and volume present in environment
+      if (volumeName2PVC.containsKey(volumeName)) {
+        pvc = volumeName2PVC.get(volumeName);
+      }
+      // when PVC is not found in environment then create new one
+      else {
+        final String uniqueName = Names.generateName(pvcNamePrefix);
+        pvc = newPVC(uniqueName, pvcAccessMode, pvcQuantity);
+        putLabel(pvc, CHE_WORKSPACE_ID_LABEL, workspaceId);
+        putLabel(pvc, CHE_VOLUME_NAME_LABEL, volumeName);
+        k8sEnv.getPersistentVolumeClaims().put(uniqueName, pvc);
+        volumeName2PVC.put(volumeName, pvc);
+      }
+
+      // binds pvc to pod and container
+      String pvcName = pvc.getMetadata().getName();
+      PodSpec podSpec = pod.getSpec();
+      Optional<io.fabric8.kubernetes.api.model.Volume> volumeOpt =
+          podSpec
+              .getVolumes()
+              .stream()
+              .filter(
+                  volume ->
+                      volume.getPersistentVolumeClaim() != null
+                          && pvcName.equals(volume.getPersistentVolumeClaim().getClaimName()))
+              .findAny();
+      io.fabric8.kubernetes.api.model.Volume podVolume;
+      if (volumeOpt.isPresent()) {
+        podVolume = volumeOpt.get();
+      } else {
+        podVolume = newVolume(pvcName, pvcName);
+        podSpec.getVolumes().add(podVolume);
+      }
+
+      container.getVolumeMounts().add(newVolumeMount(podVolume.getName(), volumePath, ""));
```

Comment:
I see we're adding volumeMounts but not removing old ones, is this intentional?

Reply:
yes, we don't remove/replace existing volumes mounts here but provision volumes mounts that are configured with Che Volumes